### PR TITLE
Managed CRI docker image cache

### PR DESCRIPTION
**Pull request by koct9i**
Add CRI Image Cache

This is SLRU cache which tracks docker images and their tags and digests.

### CODE REVIEW

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        , Config_(config)
+        , Executor_(executor)
+    { }
+
+    TFuture<void> Init() override
+    {
+        YT_LOG_DEBUG("Initializing docker image cache");
+
+        std::vector<TCallback<TFuture<TCriImageCacheEntryPtr>()>> prefetchers;
+        for (const auto& prefetchImage: Config_->PinnedImages) {
```
Comment (gritukan) [inline yt/yt/library/containers/cri/image_cache.cpp:26]: `for (const auto& x : y)` (with space on both sides of `:`)  here and below

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+            TCriImageDescriptor image{.Image = prefetchImage};
+            prefetchers.push_back(BIND(&TCriImageCache::GetImage, MakeStrong(this), image, /*authConfig*/ nullptr));
+        }
+
+        return Executor_->ListImages()
+            .Apply(BIND([=, this, this_ = MakeStrong(this)](const TCriImageApi::TRspListImagesPtr &imagesList) {
```
Comment (gritukan) [inline yt/yt/library/containers/cri/image_cache.cpp:32]: Space between `]` and `(`.

File: yt/yt/server/node/exec_node/job_environment.cpp [hunk 3]
```diff
@@ -1285,9 +1293,13 @@ IJobEnvironmentPtr CreateJobEnvironment(INodePtr configNode, IBootstrap* bootstr
...
         case EJobEnvironmentType::Cri: {
             auto criConfig = ConvertTo<TCriJobEnvironmentConfigPtr>(configNode);
+            auto executor = CreateCriExecutor(criConfig->CriExecutor);
+            auto imageCache = CreateCriImageCache(criConfig->CriImageCache, executor);
             return New<TCriJobEnvironment>(
                 criConfig,
-                bootstrap);
+                bootstrap,
+                executor,
+                imageCache);
```
Comment (k-pogorelov) [inline yt/yt/server/node/exec_node/job_environment.cpp:1302]: move

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+                        if (IsManagedImage(alias)) {
+                            auto cookie = BeginInsert(alias.Image);
+                            if (cookie.IsActive()) {
+                                DoInsertImage(cookie, protoImage);
+                            }
+                            managed = true;
+                        }
+                    }
+                    if (!managed) {
+                        YT_LOG_DEBUG("Found unmanaged docker image entry (Image: %v, ImageId: %v, ImageSize: %v)",
```
[the remark this answers is not in the document]
Comment (koct9i) [reply yt/yt/library/containers/cri/image_cache.cpp:103]: No. "unmanaged" means this cache manager should not manage it.
Its config has positive/negative filters for that.

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+        return Executor_->GetImageStatus(image)
+            .Apply(BIND([=, this, this_ = MakeStrong(this)] (const TCriImageApi::TRspImageStatusPtr& imageStatus) {
+                if (!imageStatus->has_image()) {
+                    return MakeFuture<TCriImageCacheEntryPtr>(TError("Docker image not found in cache (Image: %v)", image));
+                }
+                auto protoImage = imageStatus->image();
+                TCriImageDescriptor imageId{.Image = protoImage.id()};
+                if (!IsManagedImage(image)) {
+                    auto entry = New<TCriImageCacheEntry>(imageId, image, protoImage.size());
+                    YT_LOG_DEBUG("Unmanaged docker image found in cache (Image: %v, ImageId: %v)",
```
[the remark this answers is not in the document]
Comment (koct9i) [reply yt/yt/library/containers/cri/image_cache.cpp:151]: Yes. CRI maintains own persistent image cache - GetImageStatus lookup in that cache.
This image cache manager exposes unmanaged images if they are in cri cache.

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+        if (entry->HasAliases() || Find(entry->GetKey())) {
+            YT_LOG_DEBUG("Retain docker image (Image: %v, ImageId: %v, ImageSize: %v, LastPullTime: %v, LastUsageTime: %v)",
+                entry->ImageName(),
+                entry->ImageId(),
+                entry->GetImageSize(),
+                entry->GetLastPullTime(),
+                entry->GetLastUsageTime());
+            auto cookie = BeginInsert(entry->GetKey());
+            if (cookie.IsActive()) {
+                cookie.EndInsert(entry);
```
[the remark this answers is not in the document]
Comment (koct9i) [reply yt/yt/library/containers/cri/image_cache.cpp:359]: Overall remove is non-atomic.
Here entry still reachable by lookup - it hashed into values.
At this point entry could be linked into cache back (directly or via alias).

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+                return Executor_->GetImageStatus(image);
+            }))
+            .Apply(BIND([=, this, this_ = MakeStrong(this)] (const TCriImageApi::TRspImageStatusPtr& imageStatus) {
+                if (imageStatus->has_image() && pullPolicy != EImagePullPolicy::Always) {
+                    YT_LOG_DEBUG("Docker image already pulled into cache (Image: %v, ImageId: %v)",
+                        image,
+                        imageStatus->image().id());
+                    return MakeFuture<TCriImageApi::TRspImageStatusPtr>(imageStatus);
+                }
+                YT_LOG_DEBUG("Docker image pull started (Image: %v, Authenticated: %v)",
```
[the remark this answers is not in the document]
Comment (koct9i) [reply yt/yt/library/containers/cri/image_cache.cpp:250]: For now debug level is fine.

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+                    entry->GetLastPullTime(),
+                    entry->GetLastUsageTime());
+                TouchImage(entry);
+                return imageFuture;
+            }
+
+            YT_LOG_DEBUG("Waiting for in-flight docker image pull (Image: %v)", image);
+            return imageFuture
+                .Apply(BIND([=, this, this_ = MakeStrong(this)] (const TErrorOr<TCriImageCacheEntryPtr>& imageOrError) {
+                    if (imageOrError.IsOK()) {
```
[the remark this answers is not in the document]
Comment (k-pogorelov) [reply yt/yt/library/containers/cri/image_cache.cpp:232]: I meant move in lambda capture list too

Also lambda should be marked mutable to make move work

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+                    if (!managed) {
+                        YT_LOG_DEBUG("Found unmanaged docker image entry (Image: %v, ImageId: %v, ImageSize: %v)",
+                            protoImage.repo_tags().size() != 0 ? protoImage.repo_tags()[0] : protoImage.id(),
+                            protoImage.id(),
+                            protoImage.size());
+                    }
+                }
+                return PruneImages();
+            }))
+            .Apply(BIND([prefetchers = std::move(prefetchers)] {
```
Comment (k-pogorelov) [inline yt/yt/library/containers/cri/image_cache.cpp:111]: mutable

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+                TryRemove(image.Image, /*forbidResurrection*/ true);
+                break;
+        }
+
+        auto cookie = BeginInsert(image.Image, Config_->ImageSizeEstimation);
+
+        if (!cookie.IsActive()) {
+            auto imageFuture = cookie.GetValue();
+
+            if (auto entryOrError = imageFuture.TryGet(); entryOrError && entryOrError->IsOK()) {
```
Comment (k-pogorelov) [inline yt/yt/library/containers/cri/image_cache.cpp:218]: auto& ?

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+                    image,
+                    bool(authConfig));
+                return Executor_->PullImage(image, authConfig)
+                    .Apply(BIND([=, this, this_ = MakeStrong(this)] (const TCriImageApi::TRspPullImagePtr& result) {
+                        auto imageId = TCriImageDescriptor{.Image = result->image_ref()};
+                        return Executor_->GetImageStatus(imageId);
+                    }));
+            }))
+            .Apply(BIND([=, this, this_ = MakeStrong(this), cookie = std::move(cookie)] (
+                        const TErrorOr<TCriImageApi::TRspImageStatusPtr>& imageStatusOrError) mutable {
```
Comment (k-pogorelov) [inline yt/yt/library/containers/cri/image_cache.cpp:261]: In such cases we move { to new line

File: yt/yt/library/containers/cri/image_cache.cpp [hunk 12]
```diff
@@ -0,0 +1,482 @@
...
+                imageEntry->SetLastPullTime(pullTime);
+                imageCookie.EndInsert(std::move(imageEntry));
+            }
+            auto imageFuture = imageCookie.GetValue();
+            imageFuture.Subscribe(BIND([=, cookie = std::move(cookie)] (
+                        const TErrorOr<TCriImageCacheEntryPtr>& entryOrError) mutable
+            {
+                if (entryOrError.IsOK()) {
+                    auto imageEntry = entryOrError.Value();
+                    auto aliasEntry = New<TCriImageCacheEntry>(imageName, imageEntry);
```
Comment (k-pogorelov) [inline yt/yt/library/containers/cri/image_cache.cpp:334]: move

**Review by k-pogorelov — CHANGES_REQUESTED** [review]

**Comment by koct9i** [comment]
THis code works once in nowhere. Is it really important to optimize each and every move?

**Comment by Kontakter** [comment]
For me remarks of @k-pogorelov looks safe, why do you thinks that it could break something?

**Comment by Kontakter** [comment]
I have discuss it with @zlobober and have one more comment. We understand that remarks of @k-pogorelov are about style, but we are trying to maintain high standards of the code.

I suppose the following:
1. Fix these remarks, after that we will import and merge this PR as soon as possible.
2. We should apologise that this PR was lost for 3+ month, it is not ok :( And we are currently working on improving our SLAs for reading incoming PRs and I hope we will fix it.

File: yt/yt/library/containers/cri/image_cache.cpp (hunk 10 above)
Comment (koct9i) [reply yt/yt/library/containers/cri/image_cache.cpp:218]: where? I don't see here references

File: yt/yt/library/containers/cri/image_cache.cpp (hunk 12 above)
Comment (koct9i) [reply yt/yt/library/containers/cri/image_cache.cpp:334]: nope

File: yt/yt/library/containers/cri/image_cache.cpp (hunk 10 above)
Comment (k-pogorelov) [reply yt/yt/library/containers/cri/image_cache.cpp:218]: auto& entryOrError
No need to copy error

File: yt/yt/library/containers/cri/image_cache.cpp (hunk 12 above)
Comment (k-pogorelov) [reply yt/yt/library/containers/cri/image_cache.cpp]: Why?

File: yt/yt/library/containers/cri/image_cache.cpp (hunk 10 above)
Comment (koct9i) [reply yt/yt/library/containers/cri/image_cache.cpp:218]: TryGet returns std::optional not reference
https://github.com/ytsaurus/ytsaurus/blob/0def99575af0e0fc4fdc04fbb46dbcc77085d31a/yt/yt/core/actions/future-inl.h#L271

File: yt/yt/library/containers/cri/image_cache.cpp (hunk 12 above)
Comment (koct9i) [reply yt/yt/library/containers/cri/image_cache.cpp]: this is not subscribe-uniqe, we cannot deconstruct result here

Comment (k-pogorelov) [reply yt/yt/library/containers/cri/image_cache.cpp]: auto imageEntry = entryOrError.Value();

imageEntry was a copy, but I see that in the latest version there is a reference here

**Review by k-pogorelov — APPROVED** [review]

**Pull request closed without merging**